src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+        """
+        Obtain shifts equations from common-lines matrix
+
+        :param n_check: For each image/projection find its common-lines with
+            n_check images. If n_check is less than the total number of images,
+            a random subset of n_check images is used.
+        """
+
+        clstack = self.clmatrix
+        corrstack = self.cormatrix
+        shifts_1d = self.shifts_1d
+
+        n_img = self.n_img
+        if n_check is None:
+            n_check = n_img
+
+        n_theta = self.n_theta
+        if n_theta % 2 == 1:
+            logger.error('n_theta must be even')
+        n_theta = n_theta // 2
+
+        # Allocate variables used for shift estimation
+
+        # Based on the estimated common-lines, construct the equations
+        # for determining the 2D shift of each image. The shift equations
+        # are represented using a sparse matrix, since each row in the
+        # system contains four non-zeros (as it involves exactly four unknowns).
+        # The variables below are used to construct this sparse system.
+        # The k'th non-zero element of the equations matrix is stored as
+        # index (shift_i[k],shift_j[k]).
+
+        # Row index for sparse equations system
+        shift_i = np.zeros(4 * n_img * n_check)
+        #  Column index for sparse equations system
+        shift_j = np.zeros(4 * n_img * n_check)
+        # The coefficients of the center estimation system ordered
+        # as a single vector.
+        shift_eq = np.zeros(4 * n_img * n_check)
+        # shift_equations_map[k1,k2] is the index of the equation for
+        # the common line between images i and j.
+        shift_equations_map = np.zeros((n_img, n_img))
+        # The equation number we are currently processing
+        shift_equation_idx = 0
+        # Right hand side of the system
+        shift_b = np.zeros(n_img * (n_img - 1) // 2)
+        # Not 2*pi/n_theta, since we divided n_theta by 2 to take rays of length 2*n_r-1.
+        dtheta = np.pi / n_theta
+
+        # Go through common lines between [i, j  pairs of projections based on the
+        # corrstack values created when build the common lines matrix.
+        for i in range(n_img-1):
+
+            for j in range(i+1, n_img):
+                # skip j image without correlation
+                if corrstack[i, j] == 0:
+                    continue
+                # Create a shift equation for the projections pair (i, j).
+                idx = np.arange(4 * shift_equation_idx, 4 * shift_equation_idx + 4)
+                shift_alpha = clstack[i, j] * dtheta
+                shift_beta = clstack[j, i] * dtheta
+                shift_i[idx] = shift_equation_idx
+                shift_j[idx] = [2 * i, 2 * i + 1, 2 * j, 2 * j + 1]
+                shift_b[shift_equation_idx] = shifts_1d[i, j]
+
+                # Compute the coefficients of the current equation.
+                if shift_beta < np.pi:
```

Comment:
The latest two are different.Rearranged to make clearer.

Reply:
Which two are different? This 2×2 matrix is the same for both cases since `np.sin(shift_beta - np.pi) = -np.sin(shift_beta)` and `np.cos(shift_beta - np.pi) = -np.cos(shift_beta)`. It's just the multiplier `-1` that changes.

Reply:
Done.